### PR TITLE
refactor(BUX-411): taskmanager simplification & tasq with redis fixes

### PR DESCRIPTION
**Pull request by chris-4chain**
# Pull Request Checklist

- [x] 📖 I created my PR using provided  : [CODE_STANDARDS](https://github.com/BuxOrg/bux/blob/main/.github/CODE_STANDARDS.md)
- [x] 📖 I have read the short Code of Conduct: [CODE_OF_CONDUCT](https://github.com/BuxOrg/bux/blob/main/.github/CODE_OF_CONDUCT.md)
- [x] 🏠 I tested my changes locally.
- [x] ✅ I have provided tests for my changes.
- [x] 📝 I have used conventional commits.
- [ ] 📗 I have updated any related documentation.
- [x] 💾 PR was issued based on the Github or Jira issue.

Initially, the goal of this task was to simplify the taskmanager by fixing taskq and localCron and make some refactorization & simplification changes. All are done:
- I renamed all taskmanager.`client`-related names/filenames to more precise ones;
- Removed unnecessary taskmanager.`With` functional options;
- Simplified `Redis` configuration: `taskmanager.WithRedis` function;
- Removed our `localCron` structure, instead, the `github.com/robfig/cron/v3` - `cronService` is used directly.

After my changes, a lot of test files had to be adjusted - that's why so many files are modified. 

I tested it locally running different examples and `bux-server` with backend/frontend. Tasks were correctly scheduled. 

## TL;DR
Additionally, I tested the `bux`, running several parallel instances with Redis configuration for `taskq`. Apparently, the codebase before my changes didn't allow to run it that way. So I fixed the issue, according to [task-doc](https://taskq.uptrace.dev/guide/golang-task-queue.html#consuming-tasks), adding:

```golang
if factoryType == FactoryRedis {
  if err := q.Consumer().Start(ctx); err != nil {
	  return err
  }
}
```

After that, the `taskq` in different nodes (instances) started to communicate via Redis. 

But another problem appeared. 
Number of cronJob executions in defined period was about number of active nodes.

> For example, For `cronJob` with 30sec period and `5` active instances: The job executions among all nodes was 5.

In this PR I'm proposing a fix using `distributed redis locks` with TTL. 
- Every node has local cron enabled.
- At the beginning of a cron handler I try to lock the task like this:
```golang
tryLock = func() bool {
	boolCmd := c.options.taskq.config.Redis.SetNX(ctx, key, "1", lockTime)
	return boolCmd.Val()
}
```
- If any node didn't execute the job in `lockTime` period, the job is added to a`taskq` queue.

Diagrams for better understanding:

### Previous solution
image

### New solution
image

I tested it with 6 instances and collecting local metrics. My example task has `2sec` period and every time it was executed it sent `1`(one) to my local influxDB.

I analysed the results, grouping metrics in `2sec` aggregation window and then summing the number of points per window. 

image

image

### Note:
Of course, the solution can be developed, but, for now:
- it doesn't break the single-node architecture
- generally it allows to run app in multi-node way

In the future we should re-think whether using `taskq` is a good idea for us:
- It uses obsolete 'redis/v8' library
- It is not considered as production ready (https://github.com/vmihailenco/taskq/issues/181)

Maybe a good alternative will be [asynq](https://github.com/hibiken/asynq).

### CODE REVIEW

File: taskmanager/taskmanager.go [hunk 1]
```diff
@@ -1,4 +1,161 @@
...
+		opt(tm.options)
+	}
+
+	// Set logger if not set
+	if tm.options.logger == nil {
+		tm.options.logger = logging.GetDefaultLogger()
+	}
+
+	// Use NewRelic if it's enabled (use existing txn if found on ctx)
+	// ctx = tm.options.getTxnCtx(ctx)
```
Comment (dorzepowski) [inline taskmanager/taskmanager.go:65]: why this is commented out?

File: taskmanager/taskq.go [hunk 2]
```diff
@@ -89,76 +96,88 @@ func (c *Client) loadTaskQ() error {
...
-// registerTaskUsingTaskQ will register a new task using the TaskQ engine
-func (c *Client) registerTaskUsingTaskQ(task *Task) {
-
+// RegisterTask will register a new task using the TaskQ engine
+func (c *TaskManager) RegisterTask(name string, handler interface{}) (err error) {
 	defer func() {
-		if err := recover(); err != nil {
-			c.DebugLog(fmt.Sprintf("registering task panic: %v", err))
+		if panicErr := recover(); panicErr != nil {
+			err = fmt.Errorf(fmt.Sprintf("registering task panic: %v", panicErr))
```
Comment (dorzepowski) [inline taskmanager/taskq.go:103]: can't we use somehow zerolog logger here?

File: taskmanager/taskq.go [hunk 3]
```diff
@@ -89,76 +96,88 @@ func (c *Client) loadTaskQ() error {
 	return nil
 }
 
-// registerTaskUsingTaskQ will register a new task using the TaskQ engine
-func (c *Client) registerTaskUsingTaskQ(task *Task) {
-
+// RegisterTask will register a new task using the TaskQ engine
```
Comment (dorzepowski) [inline taskmanager/taskq.go:99]:
```suggestion
// RegisterTask will register a new task to handle asynchronously
```

File: taskmanager/taskq.go [hunk 4]
```diff
@@ -89,76 +96,88 @@ func (c *Client) loadTaskQ() error {
...
 	mutex.Lock()
+	defer mutex.Unlock()
 
-	// Check if task is already registered
-	if t := taskq.Tasks.Get(task.Name); t != nil {
-
-		// Register the task locally
-		c.options.taskq.tasks[task.Name] = t
+	if t := taskq.Tasks.Get(name); t != nil {
+		// if already registered - register the task locally
```
Comment (dorzepowski) [inline taskmanager/taskq.go:111]: maybe instead of comment, we could have log message here, like:
> task {taskname} registered in taskq engine, by different instance

File: taskmanager/taskq.go [hunk 5]
```diff
@@ -89,76 +96,88 @@ func (c *Client) loadTaskQ() error {
...
 
-	// Register and store the task
-	c.options.taskq.tasks[task.Name] = taskq.RegisterTask(convertTaskToTaskQ(task))
+	// Task message will be used to add to the queue
+	taskMessage := task.WithArgs(ctx, options.Arguments...)
 
-	mutex.Unlock()
+	// There are two ways to run a task:
+	// Option 1: Run the task immediately
+	if options.RunEveryPeriod == 0 {
```
Comment (dorzepowski) [inline taskmanager/taskq.go:141]:
```suggestion
	if runImmediately(options) {
```

File: taskmanager/taskq.go [hunk 6]
```diff
@@ -89,76 +96,88 @@ func (c *Client) loadTaskQ() error {
...
+		return c.options.taskq.queue.Add(taskMessage)
+	}
 
-	// Debugging
-	c.DebugLog(fmt.Sprintf("registering task: %s...", c.options.taskq.tasks[task.Name].Name()))
+	// Option 2: Run the task periodically using cron
```
Comment (dorzepowski) [inline taskmanager/taskq.go:145]: IMHO can be removed

File: taskmanager/taskq.go [hunk 7]
```diff
@@ -89,76 +96,88 @@ func (c *Client) loadTaskQ() error {
...
-	// Starting the execution of the task
-	c.DebugLog(fmt.Sprintf(
-		"executing task: %s... delay: %s arguments: %s",
-		options.TaskName,
-		options.Delay.String(),
-		fmt.Sprintf("%+v", options.Arguments),
-	))
+		// The runEveryPeriod should be greater than 1 second
+		if runEveryPeriod < 1*time.Second {
+			runEveryPeriod = 1 * time.Second
```
Comment (dorzepowski) [inline taskmanager/taskq.go:159]: maybe instead of default we could return an error here

File: taskmanager/taskmanager.go [hunk 8]
```diff
@@ -1,4 +1,161 @@
...
+
+type (
+
+	// TaskManager implements the Tasker interface
+	TaskManager struct {
+		options *options
+	}
+
+	// options holds all the configuration for the client
+	options struct {
```
Comment (arkadiuszos4chain) [inline taskmanager/taskmanager.go:24]: Redundant comments

File: taskmanager/taskmanager.go [hunk 9]
```diff
@@ -1,4 +1,161 @@
...
+		cronService     *cron.Cron      // Internal cron job client
+		debug           bool            // For extra logs and additional debug information
```
Comment (arkadiuszos4chain) [inline taskmanager/taskmanager.go:26]: I believe we don't need `debug` flag anymore - just use `logger.Debug()`

File: taskmanager/interface.go [hunk 10]
```diff
@@ -6,29 +6,15 @@ import (
 	taskq "github.com/vmihailenco/taskq/v3"
 )
 
-// TaskService is the task related methods
-type TaskService interface {
-	RegisterTask(task *Task) error
+// Tasker is the taskmanager client interface
+type Tasker interface {
```
Comment (dorzepowski) [inline taskmanager/interface.go:10]: Maybe better name would be `TaskEngine` or `Engine` or `Manager`

**Review by dorzepowski — CHANGES_REQUESTED** [review]

File: taskmanager/taskmanager.go [hunk 11]
```diff
@@ -1,4 +1,161 @@
...
+		queue  taskq.Queue            // Queue for TaskQ
+		tasks  map[string]*taskq.Task // Registered tasks
+	}
+)
+
+// NewTaskManager creates a new client for all TaskManager functionality
+// If no options are given, it will use local memory for the queue.
+// ctx may contain a NewRelic txn (or one will be created)
+func NewTaskManager(ctx context.Context, opts ...ClientOps) (Tasker, error) {
+	// Create a new tm with defaults
```
Comment (arkadiuszos4chain) [inline taskmanager/taskmanager.go:44]: Redundant comments

File: taskmanager/taskmanager.go [hunk 12]
```diff
@@ -1,4 +1,161 @@
...
+
+	// Load the TaskQ engine
+	if err := tm.loadTaskQ(ctx); err != nil {
+		return nil, err
+	}
+
+	// Create the cron scheduler
+	tm.ResetCron()
+
+	// Return the client
```
Comment (arkadiuszos4chain) [inline taskmanager/taskmanager.go:75]: Redundant comments

File: taskmanager/taskmanager.go [hunk 13]
```diff
@@ -1,4 +1,161 @@
...
+	return tm, nil
+}
+
+// Close will close client and any open connections
```
Comment (arkadiuszos4chain) [inline taskmanager/taskmanager.go:79]:
```suggestion
// Close the client and any open connections
```

File: taskmanager/taskq.go [hunk 14]
```diff
@@ -36,51 +49,45 @@ func DefaultTaskQConfig(name string) *taskq.QueueOptions {
...
-func convertTaskToTaskQ(task *Task) *taskq.TaskOptions {
-	return &taskq.TaskOptions{
-		Name:            task.Name,
-		Handler:         task.Handler,
-		FallbackHandler: task.FallbackHandler,
-		DeferFunc:       task.DeferFunc,
-		RetryLimit:      task.RetryLimit,
-		MinBackoff:      task.MinBackoff,
-		MaxBackoff:      task.MaxBackoff,
+	// Overwrite defaults with any set by user
```
Comment (arkadiuszos4chain) [inline taskmanager/taskq.go:53]: Redundant comments

File: taskmanager/taskq.go [hunk 15]
```diff
@@ -89,76 +96,88 @@ func (c *Client) loadTaskQ() error {
...
+		c.options.taskq.tasks[name] = taskq.RegisterTask(&taskq.TaskOptions{
+			Name:       name,
+			Handler:    handler,
+			RetryLimit: 1,
+		})
+	}
 
-		// Task already exists!
-		// c.DebugLog(fmt.Sprintf("registering task: %s... task already exists!", task.Name))
+	c.DebugLog(fmt.Sprintf("registering task: %s...", c.options.taskq.tasks[name].Name()))
```
Comment (arkadiuszos4chain) [inline taskmanager/taskq.go:122]: Use `logger.Debug()`

File: taskmanager/taskmanager.go (hunk 1 above)
Comment (chris-4chain) [reply taskmanager/taskmanager.go:65]: This had been commented out before my changes so I left this as is. I don't know why. But I'm not familiar with the `NewRelic` feature.

**Review by arkadiuszos4chain — CHANGES_REQUESTED** [review]

File: taskmanager/taskq.go (hunk 4 above)
Comment (chris-4chain) [reply taskmanager/taskq.go:111]: In standard application flow - the task registration is at the initialization stage so the re-register condition never happens. I left this option (which existed before by changes) because it broke many tests.

File: taskmanager/taskmanager.go (hunk 11 above)
Comment (chris-4chain) [reply taskmanager/taskmanager.go:44]: I don't think so, because it says that by default, the local memory is used and I think it's work to notice that. About `ctx` comment, as I said, I left it as is, because it's related to `NewRelic feature`.

File: taskmanager/taskq.go (hunk 2 above)
Comment (chris-4chain) [reply taskmanager/taskq.go:102]: I added one line after it, so it looks like this:
```golang
defer func() {
		if panicErr := recover(); panicErr != nil {
			err = fmt.Errorf(fmt.Sprintf("registering task panic: %v", panicErr))
			c.options.logger.Error().Msg(err.Error())
		}
	}()
```

I left error-returning to allow other functions to proper react to the error.

File: taskmanager/taskmanager.go (hunk 8 above)
Comment (chris-4chain) [reply taskmanager/taskmanager.go:23]: Removed.

File: taskmanager/taskmanager.go (hunk 12 above)
Comment (chris-4chain) [reply taskmanager/taskmanager.go]: Removed.

File: taskmanager/taskq.go (hunk 14 above)
Comment (chris-4chain) [reply taskmanager/taskq.go]: Removed.

File: taskmanager/taskq.go (hunk 5 above)
Comment (chris-4chain) [reply taskmanager/taskq.go]: I added a method to the `TaskRunOptions` struct. Now the line looks like this. 
```golang
if options.runImmediately() {
		return c.options.taskq.queue.Add(taskMessage)
	}
```

File: taskmanager/taskq.go (hunk 6 above)
Comment (chris-4chain) [reply taskmanager/taskq.go]: Removed. Function names are readable enough so the comments were redundant. But I left the note comment about cron `The first scheduled run`

File: taskmanager/taskq.go (hunk 7 above)
Comment (chris-4chain) [reply taskmanager/taskq.go]: Done.

File: taskmanager/interface.go (hunk 10 above)
Comment (chris-4chain) [reply taskmanager/interface.go]: I renamed to TaskEngine. Generally, do we have a naming convention for Interfaces? Like `I`Task, Task`er`, Task`Interface`? In [EffectiveGo](https://go.dev/doc/effective_go#interface-names) they mansion the one-method interfaces only and they propose `-er` suffix.

File: taskmanager/taskmanager.go (hunk 9 above)
Comment (chris-4chain) [reply taskmanager/taskmanager.go]: Done

File: taskmanager/taskq.go (hunk 15 above)
Comment (chris-4chain) [reply taskmanager/taskq.go]: Done

File: taskmanager/taskmanager.go (hunk 1 above)
Comment (dorzepowski) [reply taskmanager/taskmanager.go:61]: ok, I didn't realize that it was commented out previously

File: taskmanager/taskq.go (hunk 4 above)
Comment (dorzepowski) [reply taskmanager/taskq.go:115]: I thought that if we use Redis, then taskq.Tasks.Get(name) can return info that there is already such task registered. but maybe my assumptions was wrong

Comment (chris-4chain) [reply taskmanager/taskq.go:115]: I'll double check it. Because now, I'm not quite sure :)

File: taskmanager/interface.go (hunk 10 above)
Comment (dorzepowski) [reply taskmanager/interface.go]: No we don't have such a convention (at least I don't know about it). I suggested other names because as you said `-er` suffix is rather for single method interfaces with the similar name of the method.

File: taskmanager/taskq.go (hunk 4 above)
Comment (chris-4chain) [reply taskmanager/taskq.go:115]: I checked it. Every instance have to do it's own registration and for every instance the `taskq.Tasks.Get(name)` returns nil at the initialization stage. So I think - an additional log is not needed.

image

**Review by dorzepowski — APPROVED** [review]

**Review by arkadiuszos4chain — APPROVED** [review]
